server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationDiagnosticsService.java

```diff
@@ -0,0 +1,354 @@
+     * @return The StableMasterResult for the given localMasterHistory
+     */
+    private StableMasterResult diagnoseOnHaveNotSeenMasterRecently(MasterHistory localMasterHistory, boolean explain) {
+        // NOTE: The logic in this method will be implemented in a future PR
+        String summary = "No master has been observed recently";
+        StableMasterDetails details = StableMasterDetails.EMPTY;
+        return new StableMasterResult(CoordinationDiagnosticsStatus.RED, summary, details);
+    }
+
+    /**
+     * This returns true if this node has seen a master node within the last few seconds
+     * @return true if this node has seen a master node within the last few seconds, false otherwise
+     */
+    private boolean hasSeenMasterInHasMasterLookupTimeframe() {
+        // If there is currently a master, there's no point in looking at the history:
+        if (clusterService.state().nodes().getMasterNode() != null) {
+            return true;
+        }
+        return masterHistoryService.getLocalMasterHistory().hasSeenMasterInLastNSeconds((int) nodeHasMasterLookupTimeframe.seconds());
+    }
+
+    /*
+     * If we detect that the master has gone null 3 or more times (by default), we ask the MasterHistoryService to fetch the master
+     * history as seen from the most recent master node so that it is ready in case a health API request comes in. The request to the
+     * MasterHistoryService is made asynchronously, and populates the value that MasterHistoryService.getRemoteMasterHistory() will return.
+     * The remote master history is ordinarily returned very quickly if it is going to be returned, so the odds are very good it will be
+     * in place by the time a request for it comes in. If not, this service's status will briefly switch to yellow.
+     */
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        DiscoveryNode currentMaster = event.state().nodes().getMasterNode();
+        DiscoveryNode previousMaster = event.previousState().nodes().getMasterNode();
+        if (currentMaster == null && previousMaster != null) {
+            if (masterHistoryService.getLocalMasterHistory().hasMasterGoneNullAtLeastNTimes(unacceptableNullTransitions)) {
+                DiscoveryNode master = masterHistoryService.getLocalMasterHistory().getMostRecentNonNullMaster();
+                /*
+                 * If the most recent master was this box, there is no point in making a transport request -- we already know what this
+                 * box's view of the master history is
+                 */
+                if (master != null && clusterService.localNode().equals(master) == false) {
+                    masterHistoryService.refreshRemoteMasterHistory(master);
+                }
+            }
+        }
+    }
+
+    public record StableMasterResult(CoordinationDiagnosticsStatus status, String summary, StableMasterDetails details) {}
```

Comment:
Should this also be `CoordinationDiagonosticResult`?

Reply:
Yeah, intellij refactoring failed silently. Search and replace worked though!